### PR TITLE
Pr3 parser

### PR DESCRIPTION
**Pull request by ghukill**
### What does this PR do?

This PR continues to extend this application to parse metadata records from a completed crawl.

 ### Helpful background context

As noted throughout the README, the motivating factor for wrapping the browsertrix-crawler in a container was so that we could perform actions _after_ the crawl was to parse metadata records representing websites from that crawl.

As noted in [documentation about metadata parsing](https://github.com/MITLibraries/browsertrix-harvester/blob/pr3-parser/docs/crawl_data_parsing.md#is-this-parsing-of-metadata-records-opinionated), the creation of metadata records is an **opinionated process**.  This harvester was built to support crawling of the library websites and producing records for TIMDEX.

While this `CrawlParser` class could be extended in a number of ways to generate meaningful metadata records about non-library websites, and technically it would work for non-library websites, the additional metadata it extracts from the actual HTML content is geared towards our Library WordPress websites.

### How can a reviewer manually see the effects of these changes?

Update the docker image and run a test harvest via a Make command:

```shell
make dist-local
make test-harvest-local
```

Once the crawl is complete, a single XML should be available locally on the host machine (as the harvest was performed via a docker container) at `output/crawls/collections/homepage/homepage.xml`.

Reviewing this XML file will reveal what kind of data is parsed from the websites crawled.  

### Includes new or updated dependencies?

NO

### What are the relevant tickets?

https://mitlibraries.atlassian.net/browse/TIMX-247

### Developer

- [X] All new ENV is documented in README (or there is none)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated or is unnecessary
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: harvester/parse.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+        """Cleanup fulltext as provided by Browsertrix crawl."""
+        fulltext = fulltext.replace("\n", " ")
+        fulltext = fulltext.replace("\r", " ")
+        fulltext = fulltext.replace("\t", " ")
+        return fulltext
+
+    def _parse_fulltext_keywords(self, fulltext: str) -> str:
+        """Parse keywords from fulltext, using YAKE keyword extractor."""
+        keywords = self.kw_extractor.extract_keywords(fulltext)
+        return ",".join([keyword for keyword, _score in keywords])
```
Comment (ehanson8) [inline harvester/parse.py:307]: Do you need `_score` if you only access the first value (`keyword`)?

File: harvester/parse.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+        return tags
+
+    def _parse_fulltext(self, fulltext: str) -> str:
+        """Cleanup fulltext as provided by Browsertrix crawl."""
+        fulltext = fulltext.replace("\n", " ")
+        fulltext = fulltext.replace("\r", " ")
+        fulltext = fulltext.replace("\t", " ")
+        return fulltext
+
+    def _parse_fulltext_keywords(self, fulltext: str) -> str:
```
Comment (ehanson8) [inline harvester/parse.py:304]: Same here about `parse`, I feel like it's covering different actions/concepts here which can muddle things even if it's technically accurate, what about `extract`?

File: harvester/parse.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+            with smart_open.open(filepath, "wb") as f:
+                f.write(self.to_xml())
+        elif file_format == "tsv":
+            with smart_open.open(filepath, "wb") as f:
+                self.df.to_csv(filepath, sep="\t", index=False)
+        elif file_format == "csv":
+            with smart_open.open(filepath, "wb") as f:
+                self.df.to_csv(filepath, sep=",", index=False)
+        else:
+            msg = f"file format '{file_format}' not recognized"
```
Comment (ehanson8) [inline harvester/parse.py:441]: `message`?

**Review by ehanson8 — COMMENTED** [review]
Questions and some suggested changes

File: harvester/parse.py (hunk 1 above)
Comment (ghukill) [reply harvester/parse.py:307]: Will rework the variable naming a bit, but yes, as it's unpacking a tuple with two values.  I had considered just `_`, but thought this at least indicates to a future code reader that a "score" of some kind is available if needed.

File: harvester/cli.py [hunk 4]
```diff
@@ -37,6 +38,65 @@ def shell(ctx: click.Context) -> None:
...
+@click.pass_context
+def parse_url_content(ctx: click.Context, wacz_input_file: str, url: str) -> None:
+    """Get HTML content for a single URL.
+
+    By printing the contents, this can be redirected via stdout in the calling context.
+    """
+    parser = CrawlParser(wacz_input_file)
+    html_content = parser.get_website_content_by_url(url, decode=True)
+    # ruff: noqa: T201
+    print(html_content)  # can be redirected with stdout
```
[the remark this answers is not in the document]
Comment (ghukill) [reply harvester/cli.py:63]: Ah, interesting!  Wasn't aware of this, will look into it.

File: harvester/metadata.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+                [word.lower().strip() for word in self.FULLTEXT_KEYWORD_STOPWORDS]
+            )
+        return self._keyword_extractor
+
+
+class CrawlMetadataRecords:
+    """Class to represent and serialize structured metadata extracted from crawl."""
+
+    def __init__(self, metadata_df: pd.DataFrame) -> None:
+        self.df = metadata_df
```
Comment (ehanson8) [inline harvester/metadata.py:232]: Still would prefer `dataframe` over `df` but it's not the end of the world

File: harvester/metadata.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+            with smart_open.open(filepath, "wb") as f:
+                f.write(self.to_xml())
+        elif file_format == "tsv":
+            with smart_open.open(filepath, "wb") as f:
+                self.df.to_csv(filepath, sep="\t", index=False)
+        elif file_format == "csv":
+            with smart_open.open(filepath, "wb") as f:
+                self.df.to_csv(filepath, sep=",", index=False)
+        else:
+            msg = f"File format '{file_format}' not recognized"
```
Comment (ehanson8) [inline harvester/metadata.py:277]: `message` over `msg`?

File: harvester/wacz.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+from warcio import ArchiveIterator  # type: ignore[import]
+from warcio.recordloader import ArcWarcRecord  # type: ignore[import]
+
+from harvester.exceptions import ContextManagerRequiredError, WaczFileDoesNotExist
+
+logger = logging.getLogger(__name__)
+
+
+class WACZClient:
+    """Client of utilities to interact with a WACZ archive from a completed web crawl.
```
Comment (ehanson8) [inline harvester/wacz.py:23]: Separating this code definitely improves the overall readability

File: harvester/wacz.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+    @property
+    def wacz_archive(self) -> zipfile.ZipFile:
+        """Property that provides an open ZipFile file object of the target WACZ file.
+
+        NOTE: when accessed, this property will throw a ContextManagerRequiredError
+        exception if not operating in a context manager context.
+        """
+        if not self._via_context_manager:
+            msg = (
```
Comment (ehanson8) [inline harvester/wacz.py:89]: `message` rather than `msg`?

File: harvester/wacz.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+                "Please use this class via a context manager for operations that read "
+                "the WACZ file"
+            )
+            raise ContextManagerRequiredError(msg)
+        if self._wacz_archive is None:
+            with smart_open.open(self.wacz_filepath, "rb") as file_obj:
+                _wacz_archive = zipfile.ZipFile(io.BytesIO(file_obj.read()))
+                self._wacz_archive = _wacz_archive
+        return self._wacz_archive
```
Comment (ehanson8) [inline harvester/wacz.py:98]: While this obviously works, I'm not familiar with the private/public attribute switcheroo (forgive me for not knowing a better way to describe it) design pattern here. What's the advantage of the private attribute if it eventually becomes public?

File: harvester/wacz.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+    def _get_archive_file_obj(self, archive_filepath: str) -> IO[bytes]:
```
Comment (ehanson8) [inline harvester/wacz.py:100]: `_get_archive_file_object`?

File: harvester/wacz.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+        """Return a file-like object for a filepath in the WACZ ZipFile."""
+        try:
+            return self.wacz_archive.open(archive_filepath)
+        except KeyError as exc:
```
Comment (ehanson8) [inline harvester/wacz.py:104]: Minor but we've always done `except KeyError as e`

File: harvester/wacz.py [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+                    # filter out sitemap related pages
+                    pages_df = pages_df[~pages_df.url.str.endswith("sitemap.html")]
+                    pages_df = pages_df[~pages_df.url.str.endswith("sitemap.xml")]
+
+                    # append to list of dataframes
+                    all_pages_dfs.append(pages_df)
+
+            except WaczFileDoesNotExist:
+                msg = f"A pages file was not found in WACZ file: {pages_file}."
```
Comment (ehanson8) [inline harvester/wacz.py:132]: `message` over `msg`?

File: harvester/wacz.py [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+                logger.debug(msg)
+
+        if not all_pages_dfs:
+            msg = "Both pages.jsonl and extraPages.jsonl appear missing from WACZ file"
```
Comment (ehanson8) [inline harvester/wacz.py:136]: `message` over `msg`?

File: harvester/wacz.py [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+    def _get_cdx_df(self) -> pd.DataFrame:
+        """Create DataFrame from CDX index data.
+
+        Data returned includes more low level information about the HTTP request itself
+        like status codes, mimetypes, and specific length and offsets where request
+        response can be found in the '/archive' WARC files.
+        """
+        with self._get_archive_file_obj(
+            self.CDX_INDEX_FILEPATH
+        ) as file_obj, gzip.GzipFile(fileobj=file_obj) as decompressed_file:
```
Comment (ehanson8) [inline harvester/wacz.py:150]: `file_object`?

File: harvester/wacz.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+        NOTE: this is slower than using self.get_website_content() if the warc filename
+        and offset are known, as this needs to first perform a search via the URL to get
+        the WARC filename and offset.
+        """
+        # get warc file the URL belongs to
+        rows = self.html_websites_df[self.html_websites_df.url == url]
+        if len(rows) == 1:
+            row = rows.iloc[0]
+        else:
+            exc_msg = f"Could not find url in CDX index: {url}"
```
Comment (ehanson8) [inline harvester/wacz.py:271]: `exception_message` or `message` instead of `exc_msg`?

File: tests/test_parser.py [hunk 16]
```diff
@@ -1,91 +1,24 @@
 """tests.test_parser"""
-# ruff: noqa: SLF001, PD901
+# ruff: noqa: SLF001, PD002, PD901
```
Comment (ehanson8) [inline tests/test_parser.py:2]: Rename to `test_metadata.py`?

**Review by ehanson8 — COMMENTED** [review]
A major improvement, much easier to understand! Some questions and suggestions

File: README.md [hunk 17]
```diff
@@ -56,16 +56,16 @@ The argument `--metadata-output-file="/crawls/collections/homepage/homepage.xml"
...
 Options:
   -v, --verbose  Pass to log at debug level instead of info
   -h, --help     Show this message and exit.
 
 Commands:
   generate-metadata-records  Generate metadata records from a WACZ file.
-  harvest                    Perform a web crawl and parse structured data.
-  parse-url-content          Get HTML content for a single URL.
+  harvest                    Perform a crawl and generate metadata...
+  parse-url-content          Get HTML content for a single URL from a...
```
[the remark this answers is not in the document]
Comment (ghukill) [reply README.md:68]: Good catch.  Will attempt to reword so not truncated in CLI `--help`.

File: harvester/metadata.py [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+                if content_stripped != "":
+                    og_tag_name_friendly = og_tag_name.replace(":", "_")
+                    tags[og_tag_name_friendly] = content_stripped
+        return tags
+
+    def _remove_fulltext_whitespace(self, fulltext: str) -> str:
+        """Cleanup fulltext as provided by Browsertrix crawl."""
+        fulltext = fulltext.replace("\n", " ")
+        fulltext = fulltext.replace("\r", " ")
+        fulltext = fulltext.replace("\t", " ")
```
[the remark this answers is not in the document]
Comment (ghukill) [reply harvester/metadata.py:173]: I don't think `<str>.replace` modifies in place, but this could be changed.

For context, I think this fulltext handling will be the first thing to get improved after some rounds of testing and feedback.  So, it's pretty naive and simple now.  But will update to a chained method approach.

File: harvester/metadata.py (hunk 5 above)
Comment (ghukill) [reply harvester/metadata.py:232]: I think `_df` is a common pandas convention, but will update to `metadata_df`.

File: harvester/wacz.py (hunk 9 above)
Comment (ghukill) [reply harvester/wacz.py:98]: The advantage here, and also with the keyword extractor, is effectively caching an instance of that somewhat expensive object to create.  

The private `_wacz_archive` allows a place to store that object, so that if already loaded and saved (cached), it's available via the public, ergonomically friendly `self.wacz_archive`.

File: harvester/wacz.py (hunk 11 above)
Comment (ghukill) [reply harvester/wacz.py:104]: Nice catch, think that was some copy pasta.

File: harvester/wacz.py (hunk 9 above)
Comment (ehanson8) [reply harvester/wacz.py:98]: Got it, thanks!

File: harvester/cli.py [hunk 19]
```diff
@@ -33,10 +34,19 @@ def main(ctx: click.Context, verbose: bool) -> None:
 
 @main.command()
 @click.pass_context
-def shell(ctx: click.Context) -> None:
-    """Run a bash shell inside the docker container."""
-    # ruff: noqa: S605, S607
-    os.system("bash")
+def docker_shell(ctx: click.Context) -> None:
```
Comment (ehanson8) [inline harvester/cli.py:37]: Good rename!

**Review by ehanson8 — APPROVED** [review]
Well done, great refactor!

File: Makefile [hunk 20]
```diff
@@ -50,18 +50,23 @@ ruff-apply:
 	pipenv run ruff check --fix .
 
 # CLI commands
-shell:
-	pipenv run harvest-dockerized shell
+docker-shell:
```
Comment (jonavellecuerdo) [inline Makefile:53]: How/when is this command used vs. `test-harvest-local`?

File: docs/crawl_data_parsing.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+```
+
+Note the [WACZ](https://replayweb.page/docs/wacz-format) file here, `homepage.wacz` (where "homepage" is an arbitrary `--crawl-name`).  This file is a compressed form of the entire crawl, and is the only asset parsed by this application when generating metadata records.  This is handy for multiple reasons:
+  * this file can be saved or copied, representing the crawl in its totality
+  * parsing data from the crawl is only concerned with a single file as the entrypoint (though multiple files are later accessed inside) allowing creation of metadata records even from a remote file in S3
+
+The structure of the WACZ file is nearly identical to the uncompressed crawl results, with some minor, but important, additions:
+```text
+.
+├── archive
```
Comment (jonavellecuerdo) [inline docs/crawl_data_parsing.md:39]: To make sure I understand, everything from lines 39-59 reflect the contents of `homepage.wacz` when it is decompressed?

File: docs/crawl_data_parsing.md [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+│   ├── index.cdx.gz
+│   └── index.idx
+├── logs
+│   └── crawl-20230929152634980.log
+└── pages
+    ├── extraPages.jsonl
+    └── pages.jsonl
+```
+
+The WACZ files `pages/pages.jsonl` and `pages/extraPages.jsonl` are what create the canonical list of websites to be considered for metadata records.  While a web crawl may make hundreds of HTTP requests to gather Javascript, CSS, JSON, or any other supporting files, what we are most interested in are HTML pages (websites) that were directly included as seeds or discovered via the crawl configurations.
```
Comment (jonavellecuerdo) [inline docs/crawl_data_parsing.md:62]: When we run `make test-harvest-local`, is `homepage.wacz` decompressed in the `output/` folder?

File: harvester/cli.py [hunk 23]
```diff
@@ -31,10 +34,91 @@ def main(ctx: click.Context, verbose: bool) -> None:
...
```
Comment (jonavellecuerdo) [inline harvester/cli.py:37]: Hmm, if the Click command is not using an object from the Click context or passing an object to the Click context, is `@click.pass_context` still needed? 🤔

**Comment by jonavellecuerdo** [comment]
@ghukill Thank you so much for taking the time to carefully document important classes and methods! I would definitely need more time understand it on a much deeper level (which I aim to do via attempt to create a diagram for this application) and will probably have more feedback then. However, from a high-level perspective, I think the codebase is well-organized, documented, and it's apparent that a lot of care and thought has been put into it. :) I just have three clarifying questions above!

**Review by jonavellecuerdo — APPROVED** [review]

File: harvester/cli.py (hunk 23 above)
Comment (ghukill) [reply harvester/cli.py:37]: Nice catch!  Removing from all CLI commands that don't need it.  This _also_ surfaced a bug I hadn't caught when refactoring the Makefile commands (need to call CLI command `docker-shell` not `shell`).

File: Makefile (hunk 20 above)
Comment (ghukill) [reply Makefile:53]: The Makefile command `test-harvest-local` actually kicks off a harvest, while this `docker-shell` command merely opens a bash shell in the a docker container of the application image.  It's mostly for debugging.

File: docs/crawl_data_parsing.md (hunk 21 above)
Comment (ghukill) [reply docs/crawl_data_parsing.md:39]: That's correct!  It's just a zip file of the full results (plus a couple extra files) of a crawl.

File: docs/crawl_data_parsing.md (hunk 22 above)
Comment (ghukill) [reply docs/crawl_data_parsing.md:62]: Technically, not quite.  These [two methods](https://github.com/MITLibraries/browsertrix-harvester/blob/pr3-parser/harvester/wacz.py#L81-L105) in the `WACZClient` are important, as they support reading files from the WACZ zip archive without ever unzipping it.  

The WACZ file will be available in the host machine `./output/crawls/collections/<CRAWL_NAME>/<CRAWL_NAME>.wacz` folder, as the container path `/crawls` is mounted to host `./output/`.

And, including the `--wacz-output-file` when calling the `harvest` command, can optionally _also_ write that file somewhere else.

In summary:
  * the WACZ file is never unzipped by any normal operations of this app
  * methods exist to read these files from the zipped archive
  * the WACZ is mounted to `./output` for local dev work, but it's the `--wacz-output-file` CLI argument that would be helpful for saving that file somewhere in a normal run

File: Makefile (hunk 20 above)
Comment (jonavellecuerdo) [reply Makefile:53]: It doesn't have to be done as part of this PR, but I think it would be helpful if the README had instructions on how to use this command for debugging!
